### PR TITLE
ADD: [DEV-14285] use XLA for table model

### PR DESCRIPTION
**Pull request by benleetownsend**
> [!NOTE]
> Enables XLA for the table model with a new batch postprocessor and refactored table utils, adds optional subprocess training, and upgrades TensorFlow to 2.20 with related pipeline/model updates.
> 
> - **Table Model**:
>   - Enable XLA (`supports_xla=True`) and adjust defaults (`table_batching=False`, `batch_size=1`, `extra_data_epochs=1`).
>   - Add `TableModelBatchPostprocessor` to precompute fixed-size `row_gather`/`col_gather` and integrate via input pipeline.
>   - Refactor `finetune/base_models/bert/table_utils.py`:
>     - New `compute_masked_ragged_indices` and `build_gather_outputs` (support target seq/batch sizes; improved packing); shape/dtype hardening with `tf.ensure_shape`.
>   - Update featurizer/modeling to consume new gathers and ensure shapes; typed padding values.
> - **Training & Pipeline**:
>   - `input_pipeline`: postprocessor hooks, `bucket_by_sequence_length` API, batched Keras input signatures, progress bar supports preprocessing epochs.
>   - `model`: adapt call to use `sequence_lengths` and postprocessed inputs.
>   - `base`: cleanup clears custom gradient registry; dynamic Keras verbosity for single-step; reset pipeline state on close.
> - **TableLabeler**:
>   - Optional subprocess table training using `billiard` with stdout/stderr forwarding and progress callbacks.
> - **Tooling/Deps**:
>   - Upgrade TensorFlow to 2.20 (`Dockerfile`, `pyproject.toml`), add `billiard`; add `.pre-commit-config.yaml`.
> - **Tests**:
>   - Update/add tests for new gather/build APIs, target sizing, and training callbacks.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f062adf5d7c21b1974c662b16d2fdb3e6e3d6653. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: finetune/util/input_utils.py [hunk 1]
```diff
@@ -132,13 +138,18 @@ def wrap_tqdm(
...
         it = iter(gen())
-        desc = "Epoch {}/{}".format(
-            current_epoch + current_epoch_offset, n_epochs + total_epoch_offset
-        )
+        if current_epoch < extra_data_epochs:
+            desc = "Data Preprocessing"
+        else:
+            desc = "Epoch {}/{}".format(
+                current_epoch + current_epoch_offset + extra_data_epochs, n_epochs + total_epoch_offset - extra_data_epochs
+            )
```
[the remark this answers is not in the document]
Comment (benleetownsend) [reply finetune/util/input_utils.py:157]: Agree - Progress bar currently goes from 3/24 to 26/24

File: finetune/util/table_labeler.py [hunk 2]
```diff
@@ -754,9 +846,17 @@ def fit(
         model_inputs = self.etl.get_table_text_chunks_and_context(
             text=text, tables=tables, labels=labels
         )
-        model_inputs = self._fit_table_model(
-            model_inputs, update_hook=table_update_hook
-        )
+
+        if self.train_in_process:
+            model_inputs = self._fit_table_model_local(
```
Comment (benleetownsend) [inline finetune/util/table_labeler.py]: oops

File: finetune/util/table_labeler.py [hunk 3]
```diff
@@ -666,6 +670,128 @@ def from_table_model(cls, table_model):
         )
 
 
+def _train_table_in_subprocess(
```
Comment (benleetownsend) [inline finetune/util/table_labeler.py:673]: @sihrc If you could review this file that would be very helpful.

File: finetune/util/table_labeler.py [hunk 4]
```diff
@@ -666,6 +670,128 @@ def from_table_model(cls, table_model):
...
+        update_hook=progress_queue.put,
+    )
+    classes_queue.put(table_model.classes)
+    table_model.save(table_model_path)
+    table_model.close(update_saver=False)
+
+
+# The following helpers are celery-compatible way of forwarding stdout/stderr to the parent process.
+# Most of the credit for this goes to GPT-5.
+def setup_stdio_intercept(tmpdir: str):
```
Comment (benleetownsend) [inline finetune/util/table_labeler.py:705]: context on this bit is that celery is already doing some kind of intercepting of stdio in the worker process. So any "normal" approaches to forwarding this along seem to work locally but fail inside celery.

File: finetune/util/table_labeler.py [hunk 5]
```diff
@@ -666,6 +670,128 @@ def from_table_model(cls, table_model):
...
+
+    The open() calls will block until the parent has opened the reading end,
+    which is fine because the parent opens read ends before spawning.
+    """
+    # Open write ends (blocking) and dup2
+    stdout_w = os.open(stdout_path, os.O_WRONLY)
+    stderr_w = os.open(stderr_path, os.O_WRONLY)
+    # By convention file descriptors 1 and 2 are stdout and stderr respectively
+    os.dup2(stdout_w, 1)
+    # os.close(stdout_w)
```
Comment (sihrc) [inline finetune/util/table_labeler.py]: 👀

File: finetune/base_models/bert/model.py [hunk 6]
```diff
@@ -484,14 +485,14 @@ class TableRoBERTa(_BaseBert):
...
+    supports_xla = True
     settings = {
         **BERT_BASE_PARAMS,
         # Just incase all cells fall into the same buckets this -8 allows us to pack the batches much tighter once we add EOS and BOS
         "max_length": 2048 - 8,
-        "table_batching": True,
+        "table_batching": False,
         "chunk_tables": True,
-        "batch_size": 2,  # When table batching is true this becomes a nominal batch size. Very long docs have batch size = 1
+        "batch_size": 1,  # When table batching is true this becomes a nominal batch size. Very long docs have batch size = 1
```
Comment (madisonmay) [inline finetune/base_models/bert/model.py:495]: Batch size needs to be dropped b/c when we have to pad rather than using ragged tensors this leads to a lot of padding?

Comment (benleetownsend) [reply finetune/base_models/bert/model.py:495]: previously we used bucketed batches aka "table_batching" where most of the time we were falling into the top bucket there anyway which had a batch size of 1. - now we just do a fixed batch size of 1 with full max length. Just cuts our XLA targets in half and in this setting it didn't produce any real benefit anyway.

File: finetune/util/table_labeler.py [hunk 7]
```diff
@@ -754,9 +952,17 @@ def fit(
...
+
+        if self.train_in_process:
+            model_inputs = self._fit_table_model_subprocess(
+                model_inputs, update_hook=table_update_hook
+            )
+        else:
+            model_inputs = self._fit_table_model_local(
+                model_inputs, update_hook=table_update_hook
+            )
+
```
[the remark this answers is not in the document]
Comment (benleetownsend) [reply finetune/util/table_labeler.py:964]: okay. I think I just need a better name for this.

File: finetune/base_models/bert/table_utils.py [hunk 8]
```diff
@@ -218,39 +168,44 @@ def chunk_ragged_tensor(
...
-            tf.expand_dims(tf.expand_dims(bos_pad, 0), 0)
-        )
-        eos_pad_ragged = tf.RaggedTensor.from_tensor(
-            tf.expand_dims(tf.expand_dims(eos_pad, 0), 0)
-        )
+        bs = tf.shape(X)[0]
+        seq_len = tf.shape(X)[1]
+        mask = tf.expand_dims(
+            tf.sequence_mask(sequence_lengths, maxlen=seq_len), 0
+        )  # 1, batch, seq
```
Comment (madisonmay) [inline finetune/base_models/bert/table_utils.py:186]: You are a saint for including the shape annotations

File: finetune/base_models/bert/table_utils.py [hunk 9]
```diff
@@ -311,7 +307,58 @@ def slice_by_table_indices(
...
-    col_values.set_shape([None, None] + list(inp.shape[2:]))
+    # Casts and shapes
+    col_seq_lens = tf.cast(col_seq_lens, tf.int32)
+    pos_ids = tf.cast(pos_ids, tf.int32)
+
+    # Final adjustment to exact target sequence length when training
+    if training and (target_seq_len is not None or target_batch_size is not None):
+        seq_padding = tf.maximum(target_seq_len - tf.shape(col_values)[1], 0)
+        batch_padding = tf.maximum(target_batch_size - tf.shape(col_values)[0], 0)
+        pad0 = tf.pad(
```
Comment (madisonmay) [inline finetune/base_models/bert/table_utils.py:318]: @benleetownsend I don't totally follow the two different pad values, could you briefly explain?

File: finetune/util/input_utils.py [hunk 10]
```diff
@@ -59,57 +61,64 @@ def batch_dataset(
...
         assert isinstance(
             shapes, tuple
         ), "You cannot use table batching to predict on tables as order is not guarenteed"
 
-        return (
-            dataset.map(add_length)
-            .shuffle(500 if shuffle else 1, seed=random_seed)
+        dataset = (
+            dataset.map(add_length).shuffle(500 if shuffle else 1, seed=random_seed)
             # When we update to tf.2.13 this will change to be a method on the dataset.
```
Comment (madisonmay) [inline finetune/util/input_utils.py]: nit: comment above is out of date now

**Review by madisonmay — APPROVED** [review]
To be completely honest it's quite hard for me to review tensorflow after so long out of it.
